File: src/main/java/com/leon/bugreport/BugReportPlugin.java
package com.leon.bugreport;

import com.leon.bugreport.API.CacheCleanupListener;
import com.leon.bugreport.commands.*;
import com.leon.bugreport.expansions.BugPlaceholders;
import com.leon.bugreport.extensions.PlanHook;
import com.leon.bugreport.listeners.ItemDropEvent;
import com.leon.bugreport.listeners.ReportListener;
import com.leon.bugreport.listeners.UpdateChecker;
import org.bstats.bukkit.Metrics;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.PluginCommand;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

import java.util.*;

import static com.leon.bugreport.API.ErrorClass.logErrorMessage;
import static com.leon.bugreport.BugReportDatabase.dataSource;
import static com.leon.bugreport.BugReportDatabase.getStaticUUID;
import static com.leon.bugreport.BugReportManager.*;
import static com.leon.bugreport.gui.bugreportGUI.generateNewYML;

public class BugReportPlugin extends JavaPlugin implements Listener {
	private static BugReportPlugin instance;
	private final UpdateChecker updateChecker = new UpdateChecker(this, 110732);
	private final Set<UUID> notifiedPlayers = new HashSet<>();
	private BugReportManager reportManager;

	@NotNull
	private static List<String> getNewReports(@NotNull List<String> reports, long lastLoginTimestamp) {
		List<String> newReports = new ArrayList<>();
		for (String report : reports) {
			String[] lines = report.split("\n");
			long reportTimestampLong = 0;
			for (String line : lines) {
				if (line.startsWith("Timestamp:")) {
					reportTimestampLong = Long.parseLong(line.substring(10).trim());
				}
			}
			if (reportTimestampLong > lastLoginTimestamp) {
				newReports.add(report);
			}
		}
		return newReports;
	}

	public static BugReportPlugin getPlugin() {
		return instance;
	}

	@Override
	public void onEnable() {
		reportManager = new BugReportManager(this);

		try {
			PlanHook.getInstance().hookIntoPlan();
		} catch (NoClassDefFoundError planIsNotInstalled) {
			// Ignore catch
		}

		if (Bukkit.getPluginManager().getPlugin("PlaceholderAPI") != null) {
			new BugPlaceholders(this).register();
		}

		if (!getDataFolder().exists()) {
			if (!getDataFolder().mkdirs()) {
				plugin.getLogger().warning("Failed to create data folder.");
				logErrorMessage("Failed to create data folder.");
			}
		}

		if (getConfig().getBoolean("update-checker")) {
			updateChecker.getVersion(spigotVersion -> {
				String serverVersion = this.getDescription().getVersion();
				if (compareVersions(serverVersion, spigotVersion) < 0) {
					plugin.getLogger().info("A new version of Bug Report is available: " + spigotVersion);
				}
			});
		}

		registerCommands();
		registerListeners();

		instance = this;

		new BugReportLanguage(this);

		if (getConfig().getBoolean("metrics")) {
			new Metrics(this, 18974);
		}

		generateNewYML();
	}

	@Override
	public void onDisable() {
		bugReports.clear();

		try {
			dataSource.close();
		} catch (Exception e) {
			plugin.getLogger().warning("Failed to close database connection.");
			logErrorMessage("Failed to close database connection.");
		}

		this.getServer().getMessenger().unregisterOutgoingPluginChannel(this);
		this.getServer().getMessenger().unregisterIncomingPluginChannel(this);
	}

	@EventHandler
	public void onPlayerLeave(@NotNull PlayerQuitEvent event) {
		Player player = event.getPlayer();
		UUID playerId = player.getUniqueId();
		BugReportDatabase.setPlayerLastLoginTimestamp(playerId);
	}

	private int compareVersions(@NotNull String version1, @NotNull String version2) {
		if (debugMode) {
			plugin.getLogger().info("Comparing versions: " + version1 + " and " + version2);
		}
		String[] parts1 = version1.split("\\.");
		String[] parts2 = version2.split("\\.");

		int minLength = Math.min(parts1.length, parts2.length);

		for (int i = 0; i < minLength; i++) {
			int part1 = Integer.parseInt(parts1[i]);
			int part2 = Integer.parseInt(parts2[i]);

			if (part1 < part2) {
				return -1;
			} else if (part1 > part2) {
				return 1;
			}
		}

		if (parts1.length < parts2.length) {
			return -1;
		} else if (parts1.length > parts2.length) {
			return 1;
		}

		return 0;
	}

	@EventHandler
	public void onPlayerJoin(@NotNull PlayerJoinEvent event) {
		for (Player onlinePlayer : Bukkit.getOnlinePlayers()) {
			if (onlinePlayer.isOp() || onlinePlayer.hasPermission("bugreport.notify")) {
				UUID playerId = onlinePlayer.getUniqueId();

				if (!config.getBoolean("enableBugReportNotifications")) {
					continue;
				}

				if (notifiedPlayers.contains(playerId)) {
					return;
				}

				long lastLoginTimestamp = BugReportDatabase.getPlayerLastLoginTimestamp(playerId);

				List<String> reports = bugReports.getOrDefault(getStaticUUID(), new ArrayList<>(Collections.singletonList("DUMMY")));
				List<String> newReports = getNewReports(reports, lastLoginTimestamp);

				StringBuilder message = new StringBuilder();
				if (!newReports.isEmpty()) {
					message.append(pluginColor)
							.append(pluginTitle).append(" ")
							.append(Objects.requireNonNullElse(endingPluginTitleColor, ChatColor.GRAY))
							.append(BugReportLanguage.getValueFromLanguageFile("newReportsMessage", "You have %numReports% new reports")
									.replace("%numReports%", String.valueOf(newReports.size())))
							.append("\n");
				} else {
					message.append(pluginColor)
							.append(pluginTitle).append(" ")
							.append(Objects.requireNonNullElse(endingPluginTitleColor, ChatColor.GRAY))
							.append(BugReportLanguage.getValueFromLanguageFile("noNewReportsMessage", "You have no new reports"))
							.append("\n");
				}

				if (getConfig().getBoolean("update-checker-join")) {
					updateChecker.getVersion(spigotVersion -> {
						String serverVersion = this.getDescription().getVersion();
						if (compareVersions(serverVersion, spigotVersion) < 0) {
							message.append(pluginColor).append(pluginTitle).append(" ")
									.append(Objects.requireNonNullElse(endingPluginTitleColor, ChatColor.GRAY))
									.append("A new version of Bug Report is available:")
									.append(ChatColor.YELLOW).append(" v").append(spigotVersion);
							onlinePlayer.sendMessage(message.toString());
						}
					});
				} else {
					onlinePlayer.sendMessage(message.toString());
				}

				notifiedPlayers.add(playerId);
			}
		}
	}

	private void registerCommands() {
		if (BugReportManager.debugMode) plugin.getLogger().info("Registering commands");

		this.saveDefaultConfig();
		FileConfiguration config = this.getConfig();

		BugReportCommand bugReportCommandExecutor = new BugReportCommand(reportManager);
		UniversalTabCompleter universalTabCompleter = new UniversalTabCompleter(reportManager, config);

		Objects.requireNonNull(getCommand("buglistarchived")).setExecutor(new BugListArchivedCommand());
		Objects.requireNonNull(getCommand("buglistsettings")).setExecutor(new BugListSettingsCommand(reportManager));
		Objects.requireNonNull(getCommand("buglinkdiscord")).setExecutor(new LinkDiscordCommand(reportManager));

		PluginCommand bugReportCommand = Objects.requireNonNull(this.getCommand("bugreport"));
		bugReportCommand.setTabCompleter(universalTabCompleter);
		bugReportCommand.setExecutor(bugReportCommandExecutor);

		PluginCommand bugListCommand = Objects.requireNonNull(this.getCommand("buglist"));
		bugListCommand.setTabCompleter(universalTabCompleter);
		bugListCommand.setExecutor(new BugListCommand());
	}

	private void registerListeners() {
		getServer().getPluginManager().registerEvents(new BugReportSettings.BugReportSettingsListener(), this);
		getServer().getPluginManager().registerEvents(new BugReportManager.BugReportListener(), this);
		getServer().getPluginManager().registerEvents(new BugReportCommand(reportManager), this);
		getServer().getPluginManager().registerEvents(new ItemDropEvent(), this);
		getServer().getPluginManager().registerEvents(new ReportListener(), this);
		getServer().getPluginManager().registerEvents(this, this);

		this.getServer().getMessenger().registerOutgoingPluginChannel(this, "BungeeCord");
		this.getServer().getMessenger().registerIncomingPluginChannel(this, "BungeeCord", new com.leon.bugreport.listeners.PluginMessageListener());

		new CacheCleanupListener();
	}
}

File: src/main/java/com/bugreportmc/bugreport/BugReportPlugin.kt
package com.bugreportmc.bugreport

import com.bugreportmc.bugreport.BugReportDatabase.Companion.getPlayerLastLoginTimestamp
import com.bugreportmc.bugreport.BugReportDatabase.Companion.getStaticUUID
import com.bugreportmc.bugreport.BugReportLanguage.Companion.getValueFromLanguageFile
import com.bugreportmc.bugreport.BugReportManager.Companion.bugReports
import com.bugreportmc.bugreport.BugReportManager.Companion.debugMode
import com.bugreportmc.bugreport.BugReportManager.Companion.endingPluginTitleColor
import com.bugreportmc.bugreport.BugReportManager.Companion.pluginColor
import com.bugreportmc.bugreport.BugReportManager.Companion.pluginTitle
import com.bugreportmc.bugreport.api.CacheCleanupListener
import com.bugreportmc.bugreport.api.ErrorClass.logErrorMessage
import com.bugreportmc.bugreport.commands.*
import com.bugreportmc.bugreport.expansions.BugPlaceholders
import com.bugreportmc.bugreport.extensions.PlanHook
import com.bugreportmc.bugreport.gui.bugreportGUI
import com.bugreportmc.bugreport.listeners.ItemDropEvent
import com.bugreportmc.bugreport.listeners.PluginMessageListener
import com.bugreportmc.bugreport.listeners.ReportListener
import com.bugreportmc.bugreport.listeners.UpdateChecker
import me.clip.placeholderapi.metrics.bukkit.Metrics
import org.bukkit.Bukkit
import org.bukkit.Bukkit.getPluginManager
import org.bukkit.ChatColor
import org.bukkit.command.PluginCommand
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.plugin.Plugin
import org.bukkit.plugin.java.JavaPlugin
import java.util.*
import kotlin.math.min

class BugReportPlugin : JavaPlugin(), Listener {
	private val updateChecker: UpdateChecker = UpdateChecker(this, 110732)
	private val notifiedPlayers: MutableSet<UUID> = HashSet<UUID>()
	private lateinit var reportManager: BugReportManager

	override fun onEnable() {
		plugin = this
		reportManager = BugReportManager(this)

		try {
			PlanHook.instance?.hookIntoPlan()
		} catch (planIsNotInstalled: NoClassDefFoundError) {
			// Ignore catch
		}

		if (getPluginManager().getPlugin("PlaceholderAPI") != null) {
			BugPlaceholders(this).register()
		}

		if (!dataFolder.exists()) {
			if (!dataFolder.mkdirs()) {
				plugin.logger.warning("Failed to create data folder.")
				logErrorMessage("Failed to create data folder.")
			}
		}

		if (getConfig().getBoolean("update-checker")) {
			updateChecker.getVersion { spigotVersion: String ->
				val serverVersion: String = this.description.version
				if (compareVersions(serverVersion, spigotVersion) < 0) {
					plugin.logger.info("A new version of Bug Report is available: $spigotVersion")
				}
			}
		}

		registerCommands()
		registerListeners()

		BugReportLanguage(this)

		if (getConfig().getBoolean("metrics")) {
			Metrics(this, 18974)
		}

		bugreportGUI.generateNewYML()
	}

	override fun onDisable() {
		bugReports.forEach { (_, value) -> value.clear() }

		try {
			BugReportDatabase.dataSource.close()
		} catch (e: Exception) {
			plugin.logger.warning("Failed to close database connection.")
			logErrorMessage("Failed to close database connection.")
		}

		this.server.messenger.unregisterOutgoingPluginChannel(this)
		this.server.messenger.unregisterIncomingPluginChannel(this)
	}

	@EventHandler
	fun onPlayerLeave(event: PlayerQuitEvent) {
		val player: Player = event.player
		val playerId: UUID = player.uniqueId
		BugReportDatabase.setPlayerLastLoginTimestamp(playerId)
	}

	private fun compareVersions(version1: String, version2: String): Int {
		if (debugMode) {
			plugin.logger.info("Comparing versions: $version1 and $version2")
		}
		val parts1 = version1.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
		val parts2 = version2.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

		val minLength = min(parts1.size.toDouble(), parts2.size.toDouble()).toInt()

		for (i in 0 until minLength) {
			val part1 = parts1[i].toInt()
			val part2 = parts2[i].toInt()

			if (part1 < part2) {
				return -1
			} else if (part1 > part2) {
				return 1
			}
		}

		if (parts1.size < parts2.size) {
			return -1
		} else if (parts1.size > parts2.size) {
			return 1
		}

		return 0
	}

	@EventHandler
	fun onPlayerJoin(event: PlayerJoinEvent) {
		for (onlinePlayer in Bukkit.getOnlinePlayers()) {
			if (onlinePlayer.isOp || onlinePlayer.hasPermission("bugreport.notify")) {
				val playerId: UUID = onlinePlayer.uniqueId

				if (!config.getBoolean("enableBugReportNotifications")) {
					continue
				}

				if (notifiedPlayers.contains(playerId)) {
					return
				}

				val lastLoginTimestamp: Long = getPlayerLastLoginTimestamp(playerId)
				val reports: ArrayList<String> = bugReports.getOrDefault(
					getStaticUUID(), ArrayList<String>(
						listOf("DUMMY")
					)
				)
				val newReports = getNewReports(reports, lastLoginTimestamp)

				val message = StringBuilder()
				if (newReports.isNotEmpty()) {
					message.append(pluginColor).append(pluginTitle).append(" ").append(
						Objects.requireNonNullElse<ChatColor>(
							endingPluginTitleColor, ChatColor.GRAY
						)
					).append(
						getValueFromLanguageFile(
							"newReportsMessage", "You have %numReports% new reports"
						).replace("%numReports%", newReports.size.toString())
					).append("\n")
				} else {
					message.append(pluginColor).append(pluginTitle).append(" ").append(
						Objects.requireNonNullElse<ChatColor>(
							endingPluginTitleColor, ChatColor.GRAY
						)
					).append(
						getValueFromLanguageFile(
							"noNewReportsMessage", "You have no new reports"
						)
					).append("\n")
				}

				if (getConfig().getBoolean("update-checker-join")) {
					updateChecker.getVersion { spigotVersion: String ->
						val serverVersion: String = this.description.version
						if (compareVersions(serverVersion, spigotVersion) < 0) {
							message.append(pluginColor).append(pluginTitle).append(" ").append(
								Objects.requireNonNullElse<ChatColor>(
									endingPluginTitleColor, ChatColor.GRAY
								)
							).append("A new version of Bug Report is available:").append(ChatColor.YELLOW).append(" v")
								.append(spigotVersion)
							onlinePlayer.sendMessage(message.toString())
						}
					}
				} else {
					onlinePlayer.sendMessage(message.toString())
				}

				notifiedPlayers.add(playerId)
			}
		}
	}

	private fun registerCommands() {
		if (debugMode) {
			plugin.logger.info("Registering commands")
		}

		this.saveDefaultConfig()

		val bugReportCommandExecutor = BugReportCommand(reportManager)
		val universalTabCompleter = UniversalTabCompleter(this.getConfig())

		getCommand("buglistarchived")?.setExecutor(BugListArchivedCommand())
		getCommand("buglistsettings")?.setExecutor(BugListSettingsCommand())
		getCommand("buglinkdiscord")?.setExecutor(LinkDiscordCommand(reportManager))

		val bugReportCommand: PluginCommand = Objects.requireNonNull<PluginCommand>(this.getCommand("bugreport"))
		val bugListCommand: PluginCommand = Objects.requireNonNull<PluginCommand>(this.getCommand("buglist"))

		bugReportCommand.tabCompleter = universalTabCompleter
		bugListCommand.tabCompleter = universalTabCompleter

		bugReportCommand.setExecutor(bugReportCommandExecutor)
		bugListCommand.setExecutor(BugListCommand())
	}

	private fun registerListeners() {
		server.pluginManager.registerEvents(BugReportSettings.BugReportSettingsListener(), this)
		server.pluginManager.registerEvents(BugReportManager.BugReportListener(), this)
		server.pluginManager.registerEvents(BugReportCommand(reportManager), this)
		server.pluginManager.registerEvents(ItemDropEvent(), this)
		server.pluginManager.registerEvents(ReportListener(), this)
		server.pluginManager.registerEvents(this, this)

		server.messenger.registerOutgoingPluginChannel(this, "BungeeCord")
		server.messenger.registerIncomingPluginChannel(this, "BungeeCord", PluginMessageListener())

		CacheCleanupListener()
	}

	companion object {
		lateinit var plugin: Plugin

		private fun getNewReports(reports: ArrayList<String>, lastLoginTimestamp: Long): ArrayList<String> {
			val newReports: ArrayList<String> = ArrayList()
			for (report in reports) {
				val lines = report.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
				var reportTimestampLong: Long = 0
				for (line in lines) {
					if (line.startsWith("Timestamp:")) {
						reportTimestampLong = line.substring(10).trim { it <= ' ' }.toLong()
					}
				}
				if (reportTimestampLong > lastLoginTimestamp) {
					newReports.add(report)
				}
			}
			return newReports
		}
	}
}
